File: solutions/0/0/0/4/p0004_s0.c
/* Solution to Euler Problem 4: Largest Palindrome Product. */
#include "runner.h"

static int is_palindromic(long long number) {
    char buf[32];
    snprintf(buf, sizeof(buf), "%lld", number);
    int len = (int)strlen(buf);
    for (int i = 0, j = len - 1; i < j; i++, j--) {
        if (buf[i] != buf[j]) return 0;
    }
    return 1;
}

/* Descending pruned search for the largest palindromic product of two n-digit numbers.
   Every even-length palindrome is divisible by 11, so one factor must supply that 11:
   when `a` is not a multiple of 11 the inner factor `b` steps by 11, skipping ten of
   every eleven candidates. Iterating downward with a monotone a*b early exit prunes the
   rest. O(d^2 / 11) worst case in the count d = 9*10^(n-1) of n-digit numbers. */
const char *solve(int argc, char *argv[]) {
    static char _answer[32];
    int n = parse_int(argv[1]);

    long long max_number = 1;
    for (int i = 0; i < n; i++) max_number *= 10;
    max_number -= 1;

    long long min_number = 1;
    for (int i = 0; i < n - 1; i++) min_number *= 10;

    long long max_multiple_11 = max_number - max_number % 11;

    long long largest_palindrome = 0;
    long long a_max = 0, b_max = 0;

    for (long long a = max_number; a > min_number; a--) {
        int a_is_multiple_11 = (a % 11 == 0);
        long long b_start = a_is_multiple_11 ? max_number : max_multiple_11;
        long long b_step  = a_is_multiple_11 ? 1 : 11;

        for (long long b = b_start; b >= a; b -= b_step) {
            long long ab = a * b;
            if (ab <= largest_palindrome) break;
            if (is_palindromic(ab)) {
                a_max = a; b_max = b; largest_palindrome = ab;
                break;
            }
        }
    }

    if (show)
        printf("Largest palindrome that is a multiple of two %d-digit numbers is "
               "%lld (%lldx%lld)\n", n, largest_palindrome, a_max, b_max);

    { snprintf(_answer, sizeof _answer, "%lld", (long long)(largest_palindrome)); return _answer; }
}

File: solutions/0/0/0/4/p0004_s0.py
#!/usr/bin/env python3.14
# -*- coding: utf-8 -*-
""" Solution to Euler Problem 4: Largest Palindrome Product [Level 0]. """
from __future__ import annotations

import sys
from sys import argv, stderr
from time import perf_counter
from typing import Any


def is_palindromic(*, number: int) -> bool:
    str_number: str = str(number)
    return str_number == "".join(reversed(str_number))


def solve(*, n: int) -> int:
    largest_palindrome: int = 0
    a_max: int = 0
    b_max: int = 0
    max_number: int = 10**n - 1
    min_number: int = 10 ** (n - 1)
    max_multiple_11 = max_number - max_number % 11
    for a in range(max_number, min_number, -1):
        a_is_multiple_11 = a % 11 == 0
        for b in range(max_number if a_is_multiple_11 else max_multiple_11, a - 1, -1 if a_is_multiple_11 else -11):
            ab = a * b
            if ab <= largest_palindrome:
                break
            if is_palindromic(number=ab):
                a_max, b_max, largest_palindrome = (a, b, ab)
    if sys.argv[-1] == "--show":
        print(
            f"Largest palindrome that is a multiple of two {n}-digit numbers is {largest_palindrome} ({a_max}x{b_max})"
        )
    return largest_palindrome


def main(**kwargs: Any) -> int:
    """
    Usage: ./file.py <kwarg>... [--runs=1] [--show]
    Output: "<runs> <avg_seconds> <result>"
    """
    try:
        runs_arg: str = next((arg for arg in argv[1:] if arg.startswith("--runs=")))
        runs: int = int(runs_arg.split("=", 1)[1])
        assert runs > 0
    except (AssertionError, StopIteration, ValueError):
        runs = 1
    elapsed: list[float] = []
    result: int | None = None
    rc: int = 0
    errors: list[str] = []
    for _ in range(runs):
        _start, _result, _stop = (perf_counter(), solve(**kwargs), perf_counter())
        elapsed.append(_stop - _start)
        if result is not None and _result != result:
            errors.append(f"Expected consistent result, got {_result} previous result={result}")
        result = _result
    if result is None:
        errors.append("Expected a result, got None")
    average: float = sum(elapsed) / len(elapsed)
    if errors:
        print("\n".join(errors), file=stderr)
        rc = 1
    print(f"{runs} {average} {result}")
    return rc


if __name__ == "__main__":
    raise SystemExit(main(n=int(argv[1])))
